eclair-core/src/main/scala/fr/acinq/eclair/offer/OfferManager.scala

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright 2023 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.offer
+
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.scalacompat.Crypto.{PrivateKey, PublicKey}
+import fr.acinq.bitcoin.scalacompat.{ByteVector32, ByteVector64, Crypto}
+import fr.acinq.eclair.db.{IncomingBlindedPayment, IncomingPaymentStatus, PaymentType}
+import fr.acinq.eclair.message.{OnionMessages, Postman}
+import fr.acinq.eclair.payment.receive.MultiPartHandler
+import fr.acinq.eclair.payment.receive.MultiPartHandler.{CreateInvoiceActor, ReceivingRoute}
+import fr.acinq.eclair.payment.{Bolt12Invoice, Invoice}
+import fr.acinq.eclair.wire.protocol.CommonCodecs.{bytes32, bytes64, lengthPrefixedFeaturesCodec, millisatoshi, publicKey, timestampSecond, uint64overflow}
+import fr.acinq.eclair.wire.protocol.OfferTypes.{InvoiceRequest, InvoiceTlv, Offer}
+import fr.acinq.eclair.wire.protocol.PaymentOnion.FinalPayload
+import fr.acinq.eclair.wire.protocol.{GenericTlv, MessageOnion, OfferTypes, OnionMessagePayloadTlv, TlvStream}
+import fr.acinq.eclair.{Feature, Features, MilliSatoshi, NodeParams, TimestampMilli, TimestampSecond, randomBytes32}
+import scodec.bits.ByteVector
+import scodec.codecs._
+import scodec.{Attempt, Codec, DecodeResult}
+
+import scala.concurrent.duration.DurationInt
+import scala.util.{Success, Try}
+
+object OfferManager {
```

Comment:
Can you move the `offer` folder inside the `payment` folder?

Also, this file needs some documentation. DB-less invoice schemes should even eventually be documented in BLIPs, to allow implementations to learn from each other and find optimal strategies for what data to include in blinded paths. You should explain what you choose to include in the payment path, what trade-offs were made and how this guarantees security. Think about reviewers and future maintainers of this component, and try to make their life as easy as possible by adding comments around subtle parts of the code and on important fields/classes. You should also explain that you're deferring parts of the flow to plugins, and what those plugins' responsibilities are.

I'm particularly interested in detailing how you ensure that the blinded paths used for payments are the ones that were generated with the invoice.

Reply:
Moved and added more documentation.

The preimage is in the blinded route, so if the blinded path used for payment does not match the invoice we have no way to fulfill the payment.